stripe-core/src/main/java/com/stripe/android/core/browser/customtabs/CustomTabsManager.kt

```diff
@@ -0,0 +1,221 @@
+                .onFailure { log("OnStop: couldn't unbind, ${it.stackTraceToString()}") }
+                .onSuccess { log("OnStop: service unbound") }
+        }
+        client = null
+        connection = null
+        session = null
+    }
+
+    /**
+     * Creates or retrieves an exiting CustomTabsSession.
+     *
+     * @return a CustomTabsSession, or null if a connection to the service could not be established.
+     */
+    private fun getSession(): CustomTabsSession? {
+        when {
+            client == null -> session = null
+            session == null -> session = client?.newSession(null)
+        }
+        return session
+    }
+
+    override fun openCustomTab(
+        activity: Activity,
+        uri: Uri,
+        fallback: () -> Unit
+    ) {
+        runCatching {
+            val packageName: String? = getCustomTabsPackage(uri, activity)
+            val customTabsIntent = buildCustomTabsIntent()
+
+            // If we cant find a package name no browser that supports Custom Tabs is installed.
+            if (packageName == null) {
+                if (attemptToLaunchInNativeApp(activity, uri)) {
+                    log("uri launched on native app")
+                } else {
+                    log("Custom tabs unsupported, using fallback")
+                    fallback.invoke()
+                }
+            } else {
+                log("Opening Custom Tab with package: $packageName")
+                customTabsIntent.intent.setPackage(packageName)
+                customTabsIntent.launchUrl(activity, uri)
+            }
+        }.onFailure {
+            log("Failed to open Custom Tab, using fallback: ${it.stackTraceToString()}")
+            fallback.invoke()
+        }
+    }
+
+    @Suppress("SwallowedException")
+    private fun attemptToLaunchInNativeApp(activity: Activity, uri: Uri) = when {
+        /**
+         * Android 11 introduces a new Intent flag, FLAG_ACTIVITY_REQUIRE_NON_BROWSER, which is the recommended way
+         * to try opening a native app, as it does not require the app to declare any package manager queries.
+         */
+        Build.VERSION.SDK_INT >= VERSION_CODES.R -> try {
+            activity.startActivity(
+                Intent(Intent.ACTION_VIEW, uri)
+                    .addCategory(Intent.CATEGORY_BROWSABLE)
+                    .addFlags(Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_REQUIRE_NON_BROWSER)
+            )
+            true
+        } catch (ex: ActivityNotFoundException) {
+            false
+        }
+        /**
+         * Pre-Android 11, we need to query the package manager to see if the app is installed,
+         * which we can't do as we'd need to register all bank apps supporting app2app.
+         */
+        else -> false
+    }
+
+    private fun buildCustomTabsIntent() = CustomTabsIntent.Builder(getSession())
+        .setSendToExternalDefaultHandlerEnabled(true)
+        .setShareState(CustomTabsIntent.SHARE_STATE_OFF)
```

Comment:
We can also hide the bookmark and download buttons, if we feel extra fancy 😉

Reply:
https://github.com/stripe/stripe-android/pull/7081/commits/0c8e5d311ad1189f12acb4b61a73bc6f53457377